### ISSUE TITLE
derive(Default) suppresses warn(dead_code)

### ISSUE THREAD

**Issue by CAD97** [rust-lang/rust#98871]
Given the following code: [[playground]](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cccc52bdc184fe0561a4757b67821202)

```rust
#[derive(Default)]
struct T {}

struct U {}
```

The current output is:

```
warning: struct `U` is never constructed
 --> src/lib.rs:4:8
  |
4 | struct U {}
  |        ^
  |
  = note: `#[warn(dead_code)]` on by default
```

Ideally the output should look like:

```
warning: struct `T` is never constructed
 --> src/lib.rs:2:8
  |
2 | struct T {}
  |        ^
  |
  = note: `#[warn(dead_code)]` on by default

warning: struct `U` is never constructed
 --> src/lib.rs:4:8
  |
4 | struct U {}
  |        ^
```

After the derive, the code looks roughly like

```rust
struct T {}

#[automatically_derived]
impl Default for T {
    #[inline]
    fn default() -> T { T {} }
}

struct U {}
```

As the default implementation is both 1) itself unused and 2) `#[automatically_derived]`, ideally it should not count as a use for suppressing the `dead_code` lint on `T`.

The derived implementations for `Clone`, `Debug`, `PartialEq`, and `Hash` also show this behavior.

I seem to recall a previous change to `#[automatically_derived]` to change the interaction between the unused code lint and derived implementations, but could not find it offhand.

**Comment by matthiaskrgr** [comment]
Probably a duplicate of https://github.com/rust-lang/rust/issues/47851

**Comment by CAD97** [comment]
Definitely related but *maybe* not a duplicate; #47851 is about explicit impls whereas this is a subset for just `#[automatically_derived]` impls.

**Comment by compiler-errors** [comment]
I thought the dead code lint (or other lints?) actually mentioned they didn't do this explicitly. Like, "this lint ignores automatic derives" or something.

**Comment by compiler-errors** [comment]
Ah, unused field lint does:

```
note: `Foo` has a derived impl for the trait `Debug`, but this is intentionally ignored during dead code analysis
```

**Comment by mu001999** [comment]
After #118257, we can get expected output ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=420d55eac81c785ea2331a81138c43c6)):
```rust
#[derive(Clone, Debug)]
struct T {}
```
```
warning: struct `T` is never constructed
 --> src/lib.rs:2:8
  |
2 | struct T {}
  |        ^
  |
  = note: `T` has derived impls for the traits `Debug` and `Clone`, but these are intentionally ignored during dead code analysis
  = note: `#[warn(dead_code)]` on by default

warning: `playground` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.28s
```

But #85200 ignores `Clone` and `Debug`, not included `Default`, so we cannot get such output for `Default`.

**Comment by mu001999** [comment]
By the way, I tried to add attr `#[rustc_trivial_field_reads]` for `Default`, but it doesn't work.
